src/command.ts

```diff
@@ -0,0 +1,261 @@
+
+import { FirebaseError } from "./error";
+import { getInheritedOption } from "./utils";
+import { load } from "./rc";
+import { load as _load } from "./config";
+import * as configstore from "./configstore";
+import detectProjectRoot = require("./detectProjectRoot");
+import logger = require("./logger");
+import track = require("./track");
+import { CommanderStatic } from "commander";
+
+type ActionFunction = (...args: any[]) => Promise<any> | any;
+
+interface BeforeFunction {
+  fn: ActionFunction;
+  args: any[];
+}
+
+/**
+ * Command is a wrapper around commander to simplify our use of promise-based
+ * actions and pre-action hooks.
+ */
+export default class Command {
+  private name = "";
+  private descriptionText = "";
+  private options: any[][];
+  private actionFn: ActionFunction = (): void => {};
+  private befores: BeforeFunction[];
+  private helpText = "";
+  private client: any;
+
+  /**
+   * @param cmd the command to create.
+   */
+  constructor(private cmd: string) {
+    this.name = first(cmd.split(" ")) || "";
+    this.options = [];
+    this.befores = [];
+  }
+
+  /**
+   * Sets the description of the command.
+   * @param t a human readable description.
+   * @return the command, for chaining.
+   */
+  description(t: string): Command {
+    this.descriptionText = t;
+    return this;
+  }
+
+  /**
+   * Sets any options for the command.
+   *
+   * @example
+   *   command.option("-d, --debug", "turn on debugging", false)
+   *
+   * @param args the commander-style option definition.
+   * @return the command, for chaining.
+   */
+  option(...args: any[]): Command {
+    this.options.push(args);
+    return this;
+  }
+
+  /**
+   * Attaches a function to run before the command's action function.
+   * @param fn the function to run.
+   * @param args arguments, as an array, for the function.
+   * @return the command, for chaining.
+   */
+  before(fn: ActionFunction, ...args: any[]): Command {
+    this.befores.push({ fn: fn, args: args });
+    return this;
+  }
+
+  /**
+   * Sets the help text for the command.
+   *
+   * This text is displayed when:
+   *   - the `--help` flag is passed to the command, or
+   *   - the `help <command>` command is used.
+   *
+   * @param t the human-readable help text.
+   * @return the command, for chaining.
+   */
+  help(t: string): Command {
+    this.helpText = t;
+    return this;
+  }
+
+  /**
+   * Sets the function to be run for the command.
+   * @param fn the function to be run.
+   * @return the command, for chaining.
+   */
+  action(fn: ActionFunction): Command {
+    this.actionFn = fn;
+    return this;
+  }
+
+  /**
+   * Registers the command with the client. This is used to inisially set up
+   * all the commands and wraps their functionality with analytics and error
+   * handling.
+   * @param client the client object (from src/index.js).
+   */
+  register(client: any): void {
+    this.client = client;
+    const program: CommanderStatic = client.cli;
```

Comment:
So we define `client: any` but right away say it has `cli: CommanderStatic`.  Is it too big of a pain to give `client` a stricter type?

Reply:
I _might_ be able to? `client` comes from `src/index.js` and is... a little magical? But let me see what I can do...

Reply:
I did it. :)